tests/baselines/reference/tupleIntersectionArity.errors.txt

```diff
@@ -0,0 +1,14 @@
+tests/cases/conformance/types/rest/tupleIntersectionArity.ts(6,4): error TS2556: A spread argument must either have a tuple type or be passed to a rest parameter.
+
+
+==== tests/cases/conformance/types/rest/tupleIntersectionArity.ts (1 errors) ====
+    declare const tuple1: [number, number, number] & { _nominal_a: never };
+    declare const tuple2: [number, number, number] & [string, string, string];
+    
+    declare const f1: (...params: [number, number, number]) => void;
+    f1(...tuple1);
+    f1(...tuple2); // Error

```

Comment:
Hum, an intersection of two tuples is largely nonsensical when they're differing lengths (after all, the conflicting `length` properties intersect to `never`), however when they're the same length like this I think I'd actually prefer to see this work - it's effectively the same as the tuple where each position is an intersection of the input tuple's types at those positions.
Like if I have a `[{x: 1}]` and a `[{y: 0}]`, a `[{x: 1, y: 0}]` pretty handily satisfies both. This does extend a bit to tuples with rests, where tuples like `[{x: 1}]` and `[...{y: 0}[]]` also safely combines to [{x: 1, y: 0}]` since there are shared possible lengths between the tuples.

Reply:
All of that sounds good to me!

I do have one concern though which is that my PR currently only handles this when checking call expression arity (which is what the linked issue refers to), not when doing type-level/array spread which would still not spread properly as a tuple.

I'm not sure if changing that is in-scope for this PR or not but it might be weird to not support it everywhere.

Reply:
I've made the requested changes.

I'm not sure if you wanted tuples like `[{ x: string }, { y: string }] & [{ x: string }, ...{ z: string }[]]` to include a rest type at the end but that's what I've done as it makes the most sense to me though I'd like confirmation on that.

Numeric indexers (like `Array<T>` or `[...T[]]`) won't apply a rest argument but again please tell me if that behavior doesn't make sense.